crates/walrus-service/src/node/checkpoint.rs

```diff
@@ -0,0 +1,820 @@
+        let next_ts = if let Some(last_ts) = latest_timestamp {
+            std::cmp::max(last_ts + interval_secs, now)
+        } else {
+            now
+        };
+
+        let seconds_from_now = next_ts - now;
+        let duration = std::time::Duration::from_secs(seconds_from_now as u64);
+
+        Ok(time::Instant::now() + duration)
+    }
+
+    /// Get the timestamp of the latest checkpoint, if any.
+    pub fn get_latest_checkpoint_timestamp(
+        checkpoint_dir: &Path,
+    ) -> Result<Option<i64>, CheckpointError> {
+        let engine = Self::create_backup_engine(checkpoint_dir)?;
+        let backup_info = engine.get_backup_info();
+
+        if backup_info.is_empty() {
+            return Ok(None);
+        }
+
+        let latest_timestamp = backup_info.iter().map(|info| info.timestamp).max();
+
+        Ok(latest_timestamp)
+    }
+
+    /// Create a BackupEngine instance.
+    fn create_backup_engine(checkpoint_dir: &Path) -> Result<BackupEngine, CheckpointError> {
+        let env = Env::new().map_err(|e| CheckpointError::Other(e.into()))?;
+
+        let backup_opts = BackupEngineOptions::new(checkpoint_dir)
+            .map_err(|e| CheckpointError::Other(e.into()))?;
+
+        BackupEngine::open(&backup_opts, &env).map_err(|e| CheckpointError::Other(e.into()))
+    }
+
+    /// Delete old checkpoints to maintain the max_checkpoints limit.
+    fn purge_old_checkpoints(checkpoint_dir: &Path, max_checkpoints: usize) {
+        if max_checkpoints == 0 {
+            return;
+        }
+
+        let Ok(mut engine) = Self::create_backup_engine(checkpoint_dir) else {
+            tracing::error!("Failed to create backup engine");
+            return;
+        };
+        let result = engine
+            .purge_old_backups(max_checkpoints)
+            .map_err(|e| CheckpointError::Other(anyhow::anyhow!("Purge error: {}", e)));
+
+        match result {
+            Ok(_) => tracing::info!(
+                "purged old checkpoints, keeping {} most recent",
+                max_checkpoints
+            ),
+            Err(e) => tracing::error!(?e, "Failed to purge old checkpoints"),
+        }
+    }
+
+    /// Restore from the most recent backup.
+    pub async fn restore_latest(
+        checkpoint_dir: &Path,
+        db_path: &Path,
+        wal_dir: Option<&Path>,
+    ) -> Result<(), CheckpointError> {
+        // Create a fresh BackupEngine for this operation.
+        let mut engine = Self::create_backup_engine(checkpoint_dir)?;
+
+        let restore_opts = RestoreOptions::default();
+        let wal_path = wal_dir.unwrap_or(db_path);
+
+        tracing::info!("restoring database from latest backup");
+        engine
+            .restore_from_latest_backup(db_path, wal_path, &restore_opts)
+            .map_err(|e| CheckpointError::Other(anyhow::anyhow!("restore error: {}", e)))?;
+
+        tracing::info!("database restored successfully");
+        Ok(())
+    }
+
+    fn create_backup_impl(
```

Comment:
Consider using tracing::instrument here, instead of hand-rolling the duration tracking.